crates/polars-core/src/series/mod.rs

```diff
@@ -304,12 +304,15 @@ impl Series {
         if !dtype.is_known() || (dtype.is_primitive() && dtype == self.dtype()) {
             return Ok(self.clone());
         }
-        let ret = self.0.cast(dtype);
+        #[cfg(feature = "dtype-categorical")]
+        if let DataType::Enum(None, _) = dtype {
+            polars_bail!(ComputeError: "can not cast / initialize Enum without categories present")
+        }
```

Comment:
You're fixing a issue with the Python constructor in the cast method on the Rust side. It might work, but I don't think this is the way to go.

Reply:
@stinodego I wasn't sure about this. Should we instead put checks in both the `cast` method on the Python side and in the `Series` constructor for Enums? For example, right now you can do:

```python
pl.Series([None], dtype=pl.String).cast(pl.Enum)
# or
pl.Series([None], dtype=pl.Enum)
```
We would need to address both of these separately on the Python side, which is fine--would you suggest doing this?

Reply:
Tangential question: I'm assuming that `pl.Enum([])` should be treated equally to `pl.Enum` and that both do not have categories, is this correct?

Reply:
In retrospect, I don't think we want to remove this. We need to prevent casting of an all-null Series to an empty Enum on the rust side too--do you agree? I should regardless move the Enum check into the all-null block, since it's only relevant in that case.

Reply:
Honestly I don't know the right way to go. But your error message "can not cast / initialize" tells me something is off. An error in a casting method should not say anything about Series initialization.

But I should leave this review to @c-peters as our categorical guru :)

Reply:
We allowed `pl.Enum` for selecting all Enum columns in a context, I don't think we should allow it as a datatype. `Enum` is a datatype which should know the categories in advance, there is no good default for it.

Regarding this PR, catching it on the Rust side is not ideal. I would rather catch this on the`Python` side as this is a python thing (Class vs object) in the constructor (or other places where dtypes are valid)

Reply:
@stinodego that's actually not my error message, I re-used it from [here in `Chunkedarray::cast`](https://github.com/pola-rs/polars/blob/main/crates/polars-core/src/chunked_array/cast.rs#L237):

```rust
#[cfg(feature = "dtype-categorical")]
DataType::Enum(rev_map, ordering) => {
    let Some(rev_map) = rev_map else {
    polars_bail!(ComputeError: "can not cast / initialize Enum without categories present")
};
```

The Enum series constructor first creates a string series, and then casts to Enum, thus the code currently cannot distinguish when an Enum series is being constructed or casted to, hence the error. This is what I used to decide that we might be moving around empty Enums on the rust side, and is why I implemented it here. If we decide that empty enums are not allowable, I'm happy to put some guards over on the Python side instead. I admit I am not super familiar with the Rust API so I don't know if the equivalent of `df.select(pl.Enum)` exists. If it doesn't, then I agree that we should absolutely remove it from the cast.

As a side note, if empty Enums are truly not allowed, why is are they an `Option` in the definition?

As another side note, I wonder if we might improve the Enum constructor and avoid creating the interim String series in the first place.

Reply:
The problem is that right now we are overloading the datatype class and re-using it in the expression engine to select all columns of that type. Enum is the first type where this becomes problematic as there is no good default.  

I would be in favor of guards on the python side. But as I am not too familiar there, it would be good to discuss first.

Reply:
This has echoes of the discussion around `pl.Datetime` and using the `*` wildcard for column selection for the time zone/time unit in #13683, e.g. using `pl.Datetime(time_unit="*")` to select `pl.Datetime` objects with any time unit. See @stinodego's comment [here](https://github.com/pola-rs/polars/pull/13683#issuecomment-1889904945):

> If those data type groups were internal I would be fine with it, but they are part of the public API (which I am not too sure about actually). So they shouldn't contain invalid types just because we need that in our internal logic. No user can really do anything with a `Datetime(time_zone="*")`, right?

The analogy here is that `pl.Enum` by itself is an invalid type, so we shouldn't allow it "just because we need it in our internal logic."

There are three obvious options to me, maybe someone can think of a better one:

1) Do not allow the creation of `pl.Enum` without categories. This means no selecting by `pl.Enum` unless we used some sort of wildcard trick, which seems harder than the Datetime issu e;
2) Allow the creation of `pl.Enum` without categories, but ensure there are guards in place to prevent it from being used when it shouldn't;
3) Create some form of data type abstraction that allows for so-called "virtual" super DataTypes that are never physically instantiated, but can be used for selection. This could allow for `pl.col(pl.Datetime)` without fancy wildcard tricks as well. But I don't know how easy or hard this would be, or if it's even desirable.

Reply:
Yes, I think these are the options. I would be in favor of the third as that solves the problems long term.

Reply:
Perhaps we should use the Selector API instead of trying to solve this with the data type API?